futures-util/src/stream/stream/unzip.rs

```diff
@@ -0,0 +1,171 @@
+use crate::task::AtomicWaker;
+use alloc::sync::{Arc, Weak};
+use core::pin::Pin;
+use futures_core::stream::{FusedStream, Stream};
+use futures_core::task::{Context, Poll};
+use pin_project::{pin_project, pinned_drop};
+use std::sync::mpsc;
+
+/// SAFETY: safe because only one of two unzipped streams is guaranteed
+/// to be accessing underlying stream. This is guaranteed by mpsc. Right
+/// stream will access underlying stream only if Sender (or left stream)
+/// is dropped in which case try_recv returns disconnected error.
```

Comment:
I don't think this is correct. You can call `UnzipLeft(Right)::poll_next` and `UnzipRight(Left)::is_terminated` at the same time.

Reply:
Oh right! But I can simply maintain `is_done` boolean in `UnzipRight` and set it to `true` if I receive `Poll::Ready(None)` right?

**Or** is there a case when: `is_terminated()` returns true for underline stream but it still hasn't emitted `Poll::Ready(None)` and won't ever emit it?

If thats the case, would `UnzipLeft` be dropped if it `is_terminated`? So would it be sufficient if `is_terminated()` returned:
- `is_done` if left is **not** dropped.
- `stream.is_terminated()` if left **is** dropped, since it's now safe to call that method.